### ISSUE TITLE
Stress bar mechanic

### ISSUE THREAD

**Issue by Turtyo** [Saplings-Projects/1M_sub#55]
## Feature Description

The stress bar is another way for the player to interact with the game and adds it depth. It provides buff via and global XP bar that fills up when putting stress of enemies to 0.

## Motivation

Gameplay variety. This was voted by the design team earlier.

## Concept art / Designs

Current design of the stress bar:

- Each enemy has a stress bar, which goes from 0 to max (proper to each enemy). It starts at half the max (so for a 0->100 it starts at 50).
- The stress bar naturally goes up by a given amount each turn, proper to each each enemy.
- If the bar reaches max, the enemy does a powerful attack. After that, the bar goes back to its starting value of max/2
- The player can use cards with a "sooth" effect; this well reduce the amount of stress in the bar.
- If the player manages to reduce an enemy stress to 0 (calm): they get one XP point. The bar won't go up anymore after that point for that enemy.
- There is a global XP bar. There are ranks on the XP bar. They are spaced by a certain amount, so rank 1 might be 5XP, rank 2 is 7 **more** XP (so 12 in total), rank 3 is 10 more etc... Generally speaking the ranks are further and further apart.
- If you get to a rank, you get a buff **at the start of all the next combats**. If you get to rank 3, then you have the buff of rank 1, the buff of rank 2 and the buff of rank 3 for all future combats. Those buffs are yet to be determined, but smaller buff will be at the start and bigger buffs will require getting to a higher rank.

So if we sum up for different decks
- If you don't build mental: you have to deal with powerful attacks sometimes, but the combat should also be shorter for you because you don't try to reduce mental to 0. Having some mental cards can be good to delay the growth of the stress bar
- If you build mental: you probably won't have to deal with the powerful attack, but you take more risks by making the combat drag on, thus potentially losing more HP. You can get buffs thanks to that.

## Acceptance Criteria

- The mental bar works as described above
- The XP bar does the same, and can interact to add permanent status to the player

## Proposed Solution
- Make a `mental` class (or should this be a kind of HP bar since some things are similar, like dealing damage, setting HP...). There should probably be a way to add more functions specific to the stress bar, like setting back to half HP or stress generating each turn. Maybe by having the `mental` class extend from the health class ? We only have a `HealthComponent` I think, which is a node and not a resource, we might want to have it as a resource ?
- Add the XP bar as class of its own. It should have a way to count current XP, and defined levels of the XP bar, as well as a way to add status to the player
- There probably should be something done so that Status are kept between combats ? Might be related to #45 were we start the combat by loading the status from a permanent status list.

## Implementation steps

1.  the implementation of the mental health bar, with the attack effect when full, and a placeholder for the depletion
2.  getting XP when depletion of mental health bar and global XP bar that gives status

## Additional Context

This will probably need to be done in multiple PR because this is pretty big.

- For the look of the mental bar: a blue bar above enemy head could work nicely. In the end, it would also be good if health bar was green or red with the number alongside it (because it would look better than a simple number)
- Some enemies can have stress growing faster, and some mental effects can reduce or increase the speed at which the stress goes up
- The big attack should be custom per enemy. Doesn't mean we can't have the default be a x2 attack for example, but some enemy might have a special attack (like mini-boss and floor boss for example, or unique enemies)

**Comment by Tysterman74** [comment]
Do we want to separate out the actual XP implementation in another issue? I feel the mental mechanic stuff can be developed out first, leave a TODO stub in the mental to grant XP, then add in XP once this is done. This would help with the size of this issue.

Other than that I had a few questions:

- How did we want to display the mental bar? Will this be an actual bar that will fill up as turn progresses or just a number counter like the HP?
- Does each enemy have a different amount of stress that will go up? And if so, will there be varying amounts of stress for each enemy?
- For the powerful attack that happens, are we looking at an attack thats x2, x3, x5? Or will that be customizable per enemy?

**Comment by Turtyo** [comment]
I think we can have 2 things
- the implementation of the mental health bar, with the attack effect when full, and a placeholder for the depletion
- getting XP when depletion of mental health bar and global XP bar that gives status

- I think a blue bar above enemy head could work nicely ? In the end, I would also like health bar to be green or blue with the number alongside it (because it would look better than a simple number)
- I think we said some enemies can have stress growing faster, and that some mental effects can reduce or increase the speed at which the stress goes up
- I think the big attack should be custom per enemy. Doesn't mean we can't have the default be a x2 attack for example, but some enemy might have a special attack (like mini-boss and floor boss for example, or unique enemies)

**Comment by Tysterman74** [comment]
Sounds good! If we don't want to lose the details on this, we should update the Issue to reflect these requirements. As for the XP bar, I assume you want multiple PRs for just this one issue then? 

So two different PRs for both of these bullet points in this order:

1. the implementation of the mental health bar, with the attack effect when full, and a placeholder for the depletion
2. getting XP when depletion of mental health bar and global XP bar that gives status

**Comment by Turtyo** [comment]
Yes, I'm updating the issue.
What do you think of having the mental extend from health to already get some functions ?
Only problem I'm thinking of right now is that some functions that are activated on signals will be connected on the health being changed, and if we extend we might have some problems with the mental bar having things happen to it when it should not
But it allows to not rewrite functions, because mental in itself is just another health bar

**Comment by EChisholm** [comment]
Hello, I'm going to start working on this one, with the first order of business being giving enemies mental health bars as described. I'll try extending from the already existing health component first. My naive hope is that signals from the existing health bars won't interfere with these ones and vice versa, but we'll see... Worst case we just define new signals? I'm very much still learning Godot and GDScript, so please be patient with me. I'm ready for lots of feedback!

**Comment by Turtyo** [comment]
Don't start just yet until we have decided on a proper implementation
I was also thinking that signals from the health component might interfere if we decide to extend the mental from the health. Even if you define new signals, functions from the parent will still be connected to the signals of health
I don't know how it would work to overwrite the signal connection if mental is child of health. Is it a per function basis or a per component basis ?
This needs to be checked, maybe you could do a minimal project with such a setup to test ? ie
- Class A and B with B extending from A, both with a custom print function that says "hey I'm in A" or "Hey I'm in B", with the print of B overwriting the print of A
- A has signals, and B also, connect the custom function to their respective signals
- From a third place, emit the signals of A and B and check if everything works correctly

That will also help you with becoming more familiar with the code

**Comment by Turtyo** [comment]
Taking this issue since there has been no progress in more than 2 months and EChisholm is not available to work on it

**Comment by Turtyo** [comment]
I will not start working on the XP
